File: python/4.server.py
#!/usr/bin/env python3
"""
Simple threaded HTTP server benchmark.
Run with: python 4.server.py [server|client|both]
"""
import sys
import os
import time
import threading
import argparse
from http.server import HTTPServer, BaseHTTPRequestHandler
from socketserver import ThreadingMixIn
from concurrent.futures import ThreadPoolExecutor
import urllib.request
import urllib.error

try:
    import psutil
    HAS_PSUTIL = True
except ImportError:
    HAS_PSUTIL = False

HOST = "127.0.0.1"
PORT = 8080


def get_rss_mb():
    if HAS_PSUTIL:
        return psutil.Process(os.getpid()).memory_info().rss / (1024 * 1024)
    return 0


class HelloHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(200)
        self.send_header("Content-Type", "text/plain")
        self.end_headers()
        self.wfile.write(b"hello")

    def log_message(self, format, *args):
        pass


class ThreadedHTTPServer(ThreadingMixIn, HTTPServer):
    daemon_threads = True


def run_server():
    server = ThreadedHTTPServer((HOST, PORT), HelloHandler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nShutting down...")
        server.shutdown()


def make_request(url):
    try:
        with urllib.request.urlopen(url, timeout=10) as response:
            return response.status == 200
    except Exception:
        return False


def run_load_test(num_requests=1000, concurrency=50):
    url = f"http://{HOST}:{PORT}/"

    # Warm up
    for _ in range(10):
        make_request(url)

    rss_before = get_rss_mb()

    start_time = time.perf_counter()
    with ThreadPoolExecutor(max_workers=concurrency) as executor:
        results = list(executor.map(lambda _: make_request(url), range(num_requests)))
    elapsed = time.perf_counter() - start_time

    rss_after = get_rss_mb()
    avg_latency = (elapsed / num_requests) * 1000

    print(f"workers: {concurrency}")
    print(f"reqs: {num_requests}")
    print(f"latency: {avg_latency:.2f}ms")
    print(f"rss_delta: {rss_after - rss_before:.1f}MiB")


def run_both(num_requests=1000, concurrency=50):
    """Run server in background thread and then run load test"""
    server = ThreadedHTTPServer((HOST, PORT), HelloHandler)
    server_thread = threading.Thread(target=server.serve_forever, daemon=True)
    server_thread.start()
    time.sleep(0.3)

    run_load_test(num_requests, concurrency)

    server.shutdown()


def main():
    parser = argparse.ArgumentParser(description="HTTP Server Benchmark")
    parser.add_argument("mode", nargs="?", default="both",
                        choices=["server", "client", "both"],
                        help="Run mode: server, client, or both (default: both)")
    parser.add_argument("-n", "--requests", type=int, default=1000,
                        help="Number of requests (default: 1000)")
    parser.add_argument("-c", "--concurrency", type=int, default=50,
                        help="Concurrency level (default: 50)")

    args = parser.parse_args()

    if args.mode == "server":
        run_server()
    elif args.mode == "client":
        run_load_test(args.requests, args.concurrency)
    else:
        run_both(args.requests, args.concurrency)


if __name__ == "__main__":
    main()

File: go/4.server.go
package main

import (
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"runtime"
	"sync"
	"time"
)

const (
	HOST = "127.0.0.1"
	PORT = "8081"
)

func getRSSMiB() float64 {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	return float64(m.Sys) / (1024 * 1024)
}

func helloHandler(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusOK)
	w.Write([]byte("hello"))
}

func runServer() {
	addr := HOST + ":" + PORT
	http.HandleFunc("/", helloHandler)
	fmt.Printf("Server running on http://%s\n", addr)
	fmt.Println("Press Ctrl+C to stop")
	if err := http.ListenAndServe(addr, nil); err != nil {
		fmt.Fprintf(os.Stderr, "Server error: %v\n", err)
		os.Exit(1)
	}
}

func makeRequest(client *http.Client, url string) bool {
	resp, err := client.Get(url)
	if err != nil {
		return false
	}
	io.Copy(io.Discard, resp.Body)
	resp.Body.Close()
	return resp.StatusCode == 200
}

func runLoadTest(numRequests, concurrency int) {
	url := fmt.Sprintf("http://%s:%s/", HOST, PORT)

	client := &http.Client{
		Transport: &http.Transport{
			MaxIdleConns:        concurrency,
			MaxIdleConnsPerHost: concurrency,
			IdleConnTimeout:     30 * time.Second,
		},
		Timeout: 10 * time.Second,
	}

	// Warm up
	for i := 0; i < 10; i++ {
		makeRequest(client, url)
	}

	rssBefore := getRSSMiB()

	work := make(chan struct{}, numRequests)
	for i := 0; i < numRequests; i++ {
		work <- struct{}{}
	}
	close(work)

	start := time.Now()

	var wg sync.WaitGroup
	for i := 0; i < concurrency; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for range work {
				makeRequest(client, url)
			}
		}()
	}

	wg.Wait()
	elapsed := time.Since(start)
	rssAfter := getRSSMiB()

	avgLatency := (elapsed.Seconds() / float64(numRequests)) * 1000

	fmt.Printf("workers: %d\n", concurrency)
	fmt.Printf("reqs: %d\n", numRequests)
	fmt.Printf("latency: %.2fms\n", avgLatency)
	fmt.Printf("rss_delta: %.1fMiB\n", rssAfter-rssBefore)
}

func runBoth(numRequests, concurrency int) {
	addr := HOST + ":" + PORT
	http.HandleFunc("/", helloHandler)

	server := &http.Server{Addr: addr}
	go func() {
		server.ListenAndServe()
	}()

	time.Sleep(300 * time.Millisecond)

	runLoadTest(numRequests, concurrency)

	server.Close()
}

func main() {
	mode := flag.String("mode", "both", "Run mode: server, client, or both")
	numRequests := flag.Int("n", 1000, "Number of requests")
	concurrency := flag.Int("c", 50, "Concurrency level")
	flag.Parse()

	// Also check positional argument for mode
	if flag.NArg() > 0 {
		*mode = flag.Arg(0)
	}

	switch *mode {
	case "server":
		runServer()
	case "client":
		runLoadTest(*numRequests, *concurrency)
	case "both":
		runBoth(*numRequests, *concurrency)
	default:
		fmt.Fprintf(os.Stderr, "Unknown mode: %s\n", *mode)
		os.Exit(1)
	}
}
